calyx-backend/src/firrtl.rs

```diff
@@ -84,41 +92,108 @@ fn emit_component<F: io::Write>(
+                if !dst_set.contains(&dst_canonical_str) {
+                    // if we don't have a "is invalid" statement yet, then we have to write one.
+                    // an alternative "eager" approach would be to instantiate all possible ports
+                    // (our output ports + all children's input ports) up front.
+                    let _ = write_invalid_initialization(&asgn.dst, f);
+                    dst_set.insert(dst_canonical_str);
+                }
+                // need to write out the guard.
+                let guard_string = get_guard_string(asgn.guard.as_ref());
+                writeln!(f, "{}when {}:", SPACING.repeat(2), guard_string)?;
+                let _ = write_assignment(asgn, f, 3);
             }
-            ir::Guard::CompOp(_, _, _) => todo!(),
-            ir::Guard::Port(_) => {}
-            ir::Guard::Info(_) => todo!(),
         }
     }
 
     // Add COMPONENT END: <name> anchor
-    writeln!(f, "{}; COMPONENT END: {}", SPACING.repeat(2), comp.name)?;
+    writeln!(f, "{}; COMPONENT END: {}\n", SPACING.repeat(2), comp.name)?;
 
     Ok(())
 }
 
-// Writes a FIRRTL assignment
-fn write_assignment<F: io::Write>(
-    asgn: &ir::Assignment<ir::Nothing>,
-    f: &mut F,
-) -> CalyxResult<()> {
-    let dest_port = asgn.dst.borrow();
-    let dest_string = get_port_string(&dest_port, true);
-    let source_port = asgn.src.borrow();
-    let src_string = get_port_string(&source_port, false);
-    writeln!(f, "{}{} <= {}", SPACING.repeat(2), dest_string, src_string)?;
-    Ok(())
+// recursive function that writes the FIRRTL representation for a guard.
+fn get_guard_string(guard: &ir::Guard<ir::Nothing>) -> String {
+    match guard {
+        ir::Guard::Or(l, r) => {
+            let l_str = get_guard_string(l.as_ref());
+            let r_str = get_guard_string(r.as_ref());
+            format!("or({}, {})", l_str, r_str)
+        }
+        ir::Guard::And(l, r) => {
+            let l_str = get_guard_string(l.as_ref());
+            let r_str = get_guard_string(r.as_ref());
+            format!("and({}, {})", l_str, r_str)
+        }
+        ir::Guard::Not(g) => {
+            let g_str = get_guard_string(g);
+            format!("not({})", g_str)
+        }
+        ir::Guard::True => String::from(""),
+        ir::Guard::CompOp(op, l, r) => {
+            let l_str = get_port_string(&l.borrow(), false);
+            let r_str = get_port_string(&r.borrow(), false);
+            let op_str = match op {
+                ir::PortComp::Eq => "eq",
+                ir::PortComp::Neq => "neq",
+                ir::PortComp::Gt => "gt",
+                ir::PortComp::Lt => "lt",
+                ir::PortComp::Geq => "geq",
+                ir::PortComp::Leq => "leq",
+            };
+            format!("{}({}, {})", op_str, l_str, r_str)
+        }
+        ir::Guard::Port(port) => get_port_string(&port.borrow().clone(), false),
```

Comment:
This clone seems off to me?

Since `get_port_string` only requires a `&ir::Port` as the first argument I don't see a good reason to clone anything when you can just pass a reference to the port you have. See the invocations you have above for `CompOp`

Reply:
Thanks for the catch!! Indeed the cloning was unnecessary :sweat_smile: